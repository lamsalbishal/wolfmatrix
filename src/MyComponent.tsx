

import React, { useState, useEffect, useRef, useCallback } from 'react';
import { Text, View, TouchableOpacity, FlatList, TextInput, StyleSheet } from 'react-native';
import SearchListItem from './Interfaces'

interface Props {
    data: SearchListItem[]
}
const MyComponent: React.FC<Props> = ({ data }) => {
    const [dataSource, setDataSource] = useState<SearchListItem[]>(data);
    const [searchTerm, setSearchTerm] = useState<string>('');
    const inputRef = useRef<any>(null);
    const timeoutId = useRef<any>(null);

    const handleSearch = (e: string) => {
        setSearchTerm(e);
    };

    const searchText = useCallback((e: string) => {
        if (timeoutId.current) {
            clearTimeout(timeoutId.current);
        }
        timeoutId.current = setTimeout(() => {
            setDataSource(data.filter((item: SearchListItem) => item.name.toLowerCase().includes(e.toLowerCase())));
        }, 1000);
    }, []);

    useEffect(() => {
        searchText(searchTerm);
    }, [searchTerm, searchText]);

    const onSearchChange = (e: string) => {
        handleSearch(e);
    };


    const handleClear = () => {
        inputRef.current.clear();
        setDataSource(data)
    }

    const handleSelect = (id: string) => {
        setDataSource(prevData =>
            prevData.map(item =>
                item.id === id ? item?.status ? { ...item, status: false } : { ...item, status: true } : item
            )
        );

    };

    return (
        <View style={style.searchContainer}>
            <View style={style.searchWrapBox}>
                <TextInput
                    style={style.searchTextInput}
                    ref={inputRef}
                    onChangeText={onSearchChange}
                    value={searchTerm}
                    testID="search-input"
                />
                <TouchableOpacity onPress={handleClear}>
                    <Text>Clear</Text>
                </TouchableOpacity>
            </View>

            <FlatList
                data={dataSource}
                keyExtractor={item => item.id}
                renderItem={({ item }) => (
                    <TouchableOpacity onPress={() => handleSelect(item?.id)}>
                        <View style={style.listWrapBox}>
                            <Text testID={`item-${item.name}`} style={style.titleStyle} >{item.name}</Text>
                            <Text>{item.status ? 'Selected' : 'Not selected'}</Text>
                        </View>
                    </TouchableOpacity>
                )}
            />
        </View>
    );
};

export default MyComponent

const style = StyleSheet.create({
    searchContainer: {
        padding: 10
    },
    searchWrapBox: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
    },
    searchTextInput: {
        borderRadius: 8,
        borderWidth: 1,
        borderColor: "#000000",
        width: '85%'
    },
    listWrapBox: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        paddingTop: 8

    },
    titleStyle: {
        fontSize: 20,
        color: "#000000"
    }
})