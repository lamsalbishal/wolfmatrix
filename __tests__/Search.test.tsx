import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect';
import MyComponent from '../src/MyComponent'; // Adjust the import path as necessary
import SearchData from '../src/SearchListData';
import { jest, describe, test, expect } from '@jest/globals';



jest.useFakeTimers();

describe('MyComponent', () => {
    test('renders initial data', () => {
        const { getByTestId, getByText, toJSON } = render(<MyComponent data={SearchData} />);
        const input = getByTestId('search-input');
        expect(input).toBeTruthy();
        SearchData.forEach(item => {
            expect(getByText(item.name)).toBeTruthy();
        });
        expect(toJSON()).toMatchSnapshot();
    });

    test('filters data based on search input', async () => {

        const { getByTestId, queryByText, toJSON } = render(<MyComponent data={SearchData} />);

        const input = getByTestId('search-input');


        fireEvent.changeText(input, 'Mo');

        await act(async () => {
            jest.advanceTimersByTime(1000);
        })

        await waitFor(() => {
            expect(queryByText('Motorola')).toBeTruthy();
            expect(queryByText('Samsung')).toBeNull();

        });

        // Snapshot testing of the filtered state
        expect(toJSON()).toMatchSnapshot();
    });




});