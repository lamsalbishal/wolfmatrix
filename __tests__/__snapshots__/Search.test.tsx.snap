// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent filters data based on search input 1`] = `
<View
  style={
    {
      "padding": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      style={
        {
          "borderColor": "#000000",
          "borderRadius": 8,
          "borderWidth": 1,
          "width": "85%",
        }
      }
      testID="search-input"
      value="Mo"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Clear
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "id": "1",
          "name": "Motorola",
          "status": false,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Motorola"
            >
              Motorola
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`MyComponent renders initial data 1`] = `
<View
  style={
    {
      "padding": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      style={
        {
          "borderColor": "#000000",
          "borderRadius": 8,
          "borderWidth": 1,
          "width": "85%",
        }
      }
      testID="search-input"
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Clear
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      [
        {
          "id": "1",
          "name": "Motorola",
          "status": false,
        },
        {
          "id": "2",
          "name": "Samsung",
          "status": false,
        },
        {
          "id": "3",
          "name": "Apple",
          "status": false,
        },
        {
          "id": "4",
          "name": "OnePlus",
          "status": false,
        },
        {
          "id": "5",
          "name": "Xiaomi",
          "status": false,
        },
        {
          "id": "6",
          "name": "Vivo",
          "status": false,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Motorola"
            >
              Motorola
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Samsung"
            >
              Samsung
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Apple"
            >
              Apple
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-OnePlus"
            >
              OnePlus
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Xiaomi"
            >
              Xiaomi
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingTop": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#000000",
                  "fontSize": 20,
                }
              }
              testID="item-Vivo"
            >
              Vivo
            </Text>
            <Text>
              Not selected
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
